src/openloco/station.cpp

```diff
@@ -92,14 +230,304 @@ namespace openloco
+    }
+
     // 0x00491FE0
-    uint32_t station::calc_accepted_cargo(uint16_t ax)
+    // WARNING: this may be called with station (ebp) = -1
+    uint32_t station::calcAcceptedCargo(CargoSearchState& cargoSearchState, map_pos location, uint32_t ebx)
     {
-        registers regs;
-        regs.ax = ax;
-        regs.ebp = (int32_t)this;
-        call(0x00491FE0, regs);
-        return regs.ebx;
+        cargoSearchState.byte_112C7F2(0);
+        cargoSearchState.filter(0);
+
+        if (location.x != -1)
+        {
+            cargoSearchState.filter(ebx);
+        }
+
+        cargoSearchState.resetIndustryMap();
+
+        setStationCatchmentDisplay((1 << 0));
+
+        if (location.x != -1)
+        {
+            sub_491BF5(location, 1);
+        }
+
+        cargoSearchState.resetScores();
+        cargoSearchState.dword_112C710(0);
+
+        if (this != (station*)0xFFFFFFFF)
+        {
+            for (uint16_t i = 0; i < stationTileSize; i++)
+            {
+                auto pos = stationTiles[i];
+                auto baseZ = pos.z / 4;
+                auto tile = tilemgr::get(pos);
+                for (auto& el : tile)
+                {
+                    auto stationEl = el.as_station();
+
+                    if (stationEl != nullptr && stationEl->base_z() != baseZ && !(stationEl->flags() & (1 << 5)))
+                    {
+                        cargoSearchState.byte_112C7F2(0);
+
+                        if (stationEl->stationType() == stationType::roadStation)
+                        {
+                            auto obj = objectmgr::get<road_station_object>(stationEl->object_id());
+
+                            if (obj->flags & road_station_flags::passenger)
+                            {
+                                cargoSearchState.filter(cargoSearchState.filter() | (1 << obj->var_2C));
+                            }
+                            else if (obj->flags & road_station_flags::freight)
+                            {
+                                cargoSearchState.filter(cargoSearchState.filter() | ~(1 << obj->var_2C));
+                            }
+                        }
+                        else
+                        {
+                            cargoSearchState.filter(~0);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (cargoSearchState.filter() == 0)
+        {
+            cargoSearchState.filter(~0);
+        }
+
+        for (tile_coord_t ty = 0; ty < map_columns; ty++)
+        {
+            for (tile_coord_t tx = 0; tx < map_rows; tx++)
+            {
+                if (cargoSearchState.mapHas2(tx, ty))
+                {
+                    auto pos = map_pos(tx * tile_size, ty * tile_size);
+                    auto tile = tilemgr::get(pos);
+
+                    for (auto& el : tile)
+                    {
+                        if (!el.is_flag_4())
```

Comment:
again stick to `if (el.is_flag4()) continue;` to avoid this ridiculous indenting.